### PR TITLE
feat: make possible to change client configs via feature flag

### PR DESCRIPTION
**Pull request by gabrielfs7**
https://oat-sa.atlassian.net/browse/AUT-2082

### Solution

Adding a new interface `FeatureFlagClientConfigHandlerInterface` that enables us to add multiple configuration handles based on feature flags.

Documentation: https://github.com/oat-sa/tao-core/blob/feat/AUT-2082/feature-flag-visibility-tao-advance/models/classes/featureFlag/README.md

### TODO 

- [x] Add documentation
- [x] Validate solution with FE
- [x] Add unit tests
- [ ] Validate that installation works

### Related PRs

- https://github.com/oat-sa/extension-tao-deliver-connect/pull/116
- https://github.com/oat-sa/extension-tao-itemqti/pull/2132

### CODE REVIEW

File: models/classes/featureFlag/FeatureFlagConfigSwitcher.php [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+    /** @var common_ext_ExtensionsManager */
+    private $extensionsManager;
+
+    public function __construct(
+        ClientLibConfigRegistry $registry,
+        common_ext_ExtensionsManager $extensionsManager,
+        ContainerInterface $container
+    ) {
+        $this->registry = $registry::getRegistry();
```
Comment (hectoras) [inline models/classes/featureFlag/FeatureFlagConfigSwitcher.php:51]: Do you know if we can avoid this static method call over an instance?

I think (but I've not tried) we can declare a (maybe non-public) service in the provider like this:

```
$services->set(ClientLibConfigRegistry::class)
         ->factory([ClientLibConfigRegistry::class, 'getRegistry']);
```

and then use the standard form to pass this parameter from the provider (`service(ClientLibConfigRegistry::class)`, as it is right now), and avoid the getRegistry() call here.

**Review by hectoras — APPROVED** [review]

**Review by hectoras — APPROVED** [review]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Pull request's target is not master
- [x] Kitchen links are not exposed in the description
- [x] Commits are following conventional commits

File: models/classes/featureFlag/FeatureFlagConfigSwitcher.php (hunk 1 above)
Comment (hectoras) [reply models/classes/featureFlag/FeatureFlagConfigSwitcher.php]: 376a897ad940fa41f54909e9bfcbfb04449c8fdd :+1: :tada:

File: models/classes/featureFlag/FeatureFlagConfigSwitcher.php [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+                        $handler
+                    ]
+                )
+            )
+        );
+    }
+
+    public function removeClientConfigHandler(string $handler): void
+    {
+        $handlers = $this->registry->getMap()[self::GLOBAL_FEATURE_FLAG_CLIENT_CONFIG_HANDLERS] ?? [];
```
Comment (shpran) [inline models/classes/featureFlag/FeatureFlagConfigSwitcher.php:89]:
```suggestion
        $handlers = $this->getClientConfigHandlers();
```

**Review by shpran — APPROVED** [review]
- [x] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [ ] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Pull request's target is not `master`
- [x] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] Commits messages are meaningful
- [x] Commits are atomic
- [ ] Changelog is updated according to changes (if applicable)
- [x] Documentation is updated according to changes (if applicable)

File: models/classes/featureFlag/FeatureFlagConfigHandlerInterface.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2021 (original work) Open Assessment Technologies SA;
```
[the remark this answers is not in the document]
Comment (gabrielfs7) [reply models/classes/featureFlag/FeatureFlagConfigHandlerInterface.php]: done, thanks

File: models/classes/featureFlag/FeatureFlagConfigSwitcher.php (hunk 2 above)
Comment (gabrielfs7) [reply models/classes/featureFlag/FeatureFlagConfigSwitcher.php]: logic no longer required. Thanks

File: models/classes/featureFlag/FeatureFlagConfigSwitcher.php [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+            }
+
+            /** @var FeatureFlagConfigHandlerInterface $handler */
+            $handler = $this->container->get($handlerId);
+
+            if (!$handler instanceof FeatureFlagConfigHandlerInterface) {
+                continue;
+            }
+
+            $config = $handler($config);
```
[the remark this answers is not in the document]
Comment (gabrielfs7) [reply models/classes/featureFlag/FeatureFlagConfigSwitcher.php]: Done

**Review by jsconan — COMMENTED** [review]
I did only a high-level review, looking at the logic mainly. It looks great!

Unfortunately, the PR was merged before I complete the review :D
I had a few comments, but no blocker.

File: test/unit/models/classes/featureFlag/Repository/FeatureFlagRepositoryTest.php [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+    public function testGet(): void
+    {
+        $this->markTestIncomplete();
+    }
+
+    public function testClearCache(): void
+    {
+        $this->markTestIncomplete();
+    }
```
[the remark this answers is not in the document]
Comment (gabrielfs7) [reply test/unit/models/classes/featureFlag/Repository/FeatureFlagRepositoryTest.php:69]: I am adding in another PR. Thanks for the heads up.

**Pull request merged**